Write docs for new Pallas collective APIs. (#19420)
google/jax: docs/pallas/tpu/collectives.rst

Pallas Collectives
===================

> Pallas now has support for manual DMA (between memories of the same chip) and remote DMA (between memories of different chips) support. It also has a more advanced API for nesting pipelines of compute overlapped collectives.

Reviewer: ```suggestion
Pallas has support for manual DMA (between memories of the same chip) and remote DMA (between memories of different chips) support. It also has an API for nesting pipelines of compute overlapped collectives.
```

Things like "now" in docs get stale, and it's hard to remember to update things. I'd save that language for announcements / changelog entries, and instead write docs in a 'timeless' way from the beginning.

  from jax.experimental.pallas import tpu as pltpu

  def move_data(src_ref, dst_ref, sem, send_sem, recv_sem):
    local_copy = pltpu.async_copy(src_ref, dst_ref.at[0:8], sem)
    local_copy.wait()

    # Should do a barrier the first time we communicate
    # with another chip.
    barrier_sem = pltpu.get_barrier_semaphore()
    pltpu.semaphore_signal(barrier_sem, device_id=neighbor)
    pltpu.semaphore_wait(barrier_sem)

    dev_id = pltpu.device_id()
    other_dev_id = 1 - dev_id
    remote_copy = pltpu.async_remote_copy(
      src_ref, dst_ref.at[0:8], send_sem, recv_sem, other_dev_id,
      device_id_type=pltpu.DeviceIdType.LOGICAL,
    )
    # remote_copy.wait() for both.
    remote_copy.wait_send()
    remote_copy.wait_recv()

    # Both APIs have make_ variants for building a descriptor without starting
    # it. This allows us to start a DMA in one grid iteration and wait for it
    # in another.
    local_copy = pltpu.make_async_copy(src_ref, dst_ref.at[0:8], sem)
    local_copy.start()
    local_copy.wait()
  
  src_data = pltpu.VMEM((8,128), jnp.float32)
  dst_data = pltpu.VMEM((16,128), jnp.float32)
  semaphore = pltpu.SemaphoreType.DMA

>   # Inside a kernel:

Reviewer 2: Are both this code and the call below runnable in the same context? Make sure the example runs, as stated.

  pltpu.run_scoped(
    move_data, src_data, dst_data, *((semaphore,) * 3)
  )

  # Or alternatively to use data that is live across iterations:
  pl.pallas_call(
      move_data,
      grid_spec=pltpu.PrefetchScalarGridSpec(
          num_scalar_prefetch=0,
          # Could also use other memory spaces and input/outputs.
          scratch_shapes=[src_data, dst_data, semaphore],
      ),
  )()

.. warning::

>    Care must be taken when using DMAs to write to inputs or outputs as the reads and writes to these are pipelined with the execution of the body of your kernel. The pipeline double buffering will cause undefined behavior if you try to carry out a DMA across grid iterations where the block windows/indices of the operands are changing. Modifying inputs is also generally not supported as there is no mechanism for the writes to make it to the source buffer from the current ref of the current block.

Reviewer 2: What, exactly is the contract?

Perhaps something like:
"Manual DMAs to the inputs or outputs of a pipeline is undefined behavior: there are no guarantees about the ordering of manual DMAs with respect to the pipeline's reads and writes."

Custom Pipelines
-----------------

Pipeline/Grid Recap
~~~~~~~~~~~~~~~~~~~~

> In :doc:`../quickstart` we saw how when we write a kernel like this:

Reviewer 2: Perhaps:
Recall the `matmul` example from :doc:`../quickstart`:

.. code-block:: python

  def matmul_kernel(x_ref, y_ref, z_ref):
    z_ref[...] = x_ref[...] @ y_ref[...]

  def matmul(x: jax.Array, y: jax.Array):
    return pl.pallas_call(
      matmul_kernel,
      out_shape=jax.ShapeDtypeStruct((x.shape[0], y.shape[1]), x.dtype),
      grid=(2, 2),
      in_specs=[
        pl.BlockSpec(lambda i, j: (i, 0), (x.shape[0] // 2, x.shape[1])),
        pl.BlockSpec(lambda i, j: (0, j), (y.shape[0], y.shape[1] // 2))
      ],
      out_specs=pl.BlockSpec(
        lambda i, j: (i, j), (x.shape[0] // 2, y.shape[1] // 2)
      )
    )(x, y)
  k1, k2 = jax.random.split(jax.random.PRNGKey(0))
  x = jax.random.normal(k1, (1024, 1024))
  y = jax.random.normal(k2, (1024, 1024))
  z = matmul(x, y)
  np.testing.assert_allclose(z, x @ y)

> We do a series of pipelined nested loops over the grid indices, slicing and prefetching the inputs for the next iteration and writing the outputs of the previous iteration as we are running the current one.

Reviewer 2: Perhaps:
"The `pallas_call` performs ..." instead of "We do". You didn't, Pallas did.

.. image:: ../../_static/pallas/BlockSpec.png
   :align: center
   :alt: Kernel block spec visualization.

Problem: Collective Matmuls & Compute vs. Comms Windows
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> The benefits of this declarative approach to writing kernels comes at the cost of coupling the windows you do compute on with the windows you do communication on.

Reviewer 2: I think this section needs work. I'm trying to parse the motivation and having trouble.

Initially when reading this I thought: "ah, he's motivating nested pipelines". But the first part of the text is motivating collective matmuls.

Perhaps be clearer about that: (a) first motivate collective matmuls, perhaps with a picture and then (b) show why nested pipelines are needed to do it?

.. image:: ../../_static/pallas/ag_lhs_collective_matmul.png
   :align: center
   :alt: Matmul left hand side overlapped all-gather diagram.

For optimal utilization we'd want to make sure we have enough compute to completely hide all the transfers.

.. code-block:: python

  num_flops = (2 * m * n * k) / chips
  # Assume 2 bytes per element.
  data = 2 * m * k
  is_compute_bound = (num_flops / data) > (flops / bw)
  is_compute_bound = n > chips * (flops / bw)

Depending on the number of chips and their arithmetic intensity, this can put quite a bit of upwards pressure on the size of n. In some cases that might be fine. For certain complex kernels it might not be optimal or even possible to allocate blocks big enough to make this equation hold.

Solution: Custom Pipelines == Nested Pallas Calls!
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Ideally we can use Pallas' templating capabilities to turn any kernel into one that overlaps some sort of communication pattern. Otherwise we'd be rewriting every kernel for every collective we want to overlap it with. This is where the idea for nested grids came from.

.. note::

>    The :code:`emit_pipeline` API is implemented entirely using public Pallas APIs, a testament to the power of Pallas meta-programming.

Reviewer: ```suggestion
   The :code:`emit_pipeline` API is implemented entirely using public Pallas APIs, an illustration of Pallas meta-programming.
```

Don't want to sound self-aggrandizing :P

      out_specs=pl.BlockSpec(memory_space=pltpu.TPUMemorySpace.ANY),
  )(x, y)

Nested Pipeline with Output Accumulation
*****************************************

.. code-block:: python

  def matmul_kernel(x_ref, y_ref, z_ref):

    def parent_matmul_kernel(should_accumulate_out):
      pltpu.emit_pipeline(
          existing_matmul_kernel,
          grid=(4, 4, 4),
          in_specs=[
            pl.BlockSpec(lambda i, j, k: (i, k), (128, 128)),
            pl.BlockSpec(lambda i, j, k: (k, j), (128, 128)),
        ],
        out_specs=pl.BlockSpec(lambda i, j, k: (i, j), (128, 128)),
        # Prefix-pytree of out_specs.
        should_accumulate_out=should_accumulate_out,
      )(x_ref, y_ref, z_ref)

    emit_pipeline(False)
    emit_pipeline(True)

  z = pl.pallas_call(
      matmul_kernel,
      out_shape=jax.ShapeDtypeStruct((512, 512), jnp.float32),
      in_specs=[
          pl.BlockSpec(memory_space=pltpu.TPUMemorySpace.ANY),
          pl.BlockSpec(memory_space=pltpu.TPUMemorySpace.ANY),
      ],
      out_specs=pl.BlockSpec(memory_space=pltpu.TPUMemorySpace.ANY),
  )(x, y)

Fusing Nested Pipelines (Realistic AG Matmul)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> To solve the fact that each child pipeline introduces its own bubbles at their start and end, we must add a layer of complexity via prologue and epilogue callbacks. We danced around with more restrictive but simpler ways of expressing this, but ultimately the behavior here will be very specific to the collective pattern being implemented and you might want to prefetch or wait on pipeline arguments differently based on which ones are participating in the collective.

Author: reworded

For pipeline fusion to work we also need to allocate things at the parent grid level. :code:`emit_pipeline_with_allocations` makes that simple.

> Here is a full realistic example:

Reviewer 2: The jump in complexity from the previous toy pipelines that had no communication is very large here. Can we ease into it?
